src/main/java/fi/aalto/cs/intellij/common/BuildInfo.java

```diff
@@ -0,0 +1,38 @@
+package fi.aalto.cs.intellij.common;
+
+import java.util.Properties;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Holds information that was gathered during the build of the program (such as version).
+ * Normally, this class should be accessed via {@code BuildInfo.SingletonUtility.INSTANCE}.
+ */
+public class BuildInfo {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BuildInfo.class);
+  private static final String RESOURCE_NAME = "build-info.properties";
+
+  public final Version version;
+
+  BuildInfo(@Nullable Properties props) {
+    Version ver = Version.fromProperties(props);
+    if (ver == null) {
+      ver = new Version(0,0,0);
```

Comment:
I'm somehow unsure about this, why "0, 0, 0"?

Reply:
I'd suggest that possibly better composition would be achieved, if maybe `Version` class itself could tell if the provided data is invalid and reject such request already on the instance initialization stage

Reply:
(Sorry for a long post... :P)

I think your point is very valid, but please let me explain my thought on this. :)

What I have tried to achieve (with pretty much all the code in this PR), is that I could read the version info of the software like this
```java
int versionMajor = BuildInfo.INSTANCE.version.major;
// or BuildInfo.getInstance().getVersion().getMajor() or whatever; that doesn't matter now
```
without (in the context of this "client code")
* having to handle any exceptions,
* being unsure if `INSTANCE` or `version` could be null,
* being unsure if version information is initialized yet or not.

Simply: just give me the version of this plugin where I need it.

I think it is totally acceptable to just assume that the version info is always available at run time. Why? I don't say it won't happen (it probably will happen at least once before we are done with thorough integration testing). But if it's *not* available, it's because there is **a bug** (either in our code or in our build configuration that failed to create a valid properties file and make it accessible at run time, that is, include it in the JAR).

When there is bug, the program fails, right? (That is the definition of bug, I guess.) Here the failure is the softest one (just write error log and use "0.0.0" as version, allowing the program execution go on even though in a state where it does not work as it should). The other option is to throw a RuntimeException and let the plugin (and IntelliJ?) die, which, IMO, is not a bad option either. What is common to both of these, neither of them should ever happen in production (but bugs still happen sometimes, of course). *

So why handle this issue here and not in `Version` class. One reason is that it might be totally legit to call `Version.fromString(String)` with an invalid argument. Because that argument may, for example, come from a course config file where it tells the compatible version of the plugin to be used with that file. And that string can be anything (perhaps it's mistyped by some teacher). In that case, we don't want an error log to be written (it's not an error of our program). We probably just abort loading the course and inform the user that the file they tried to use had invalid compatibility information (or just show a warning or whatever we'll decide to do). The point is, the situation is totally different and, thus, is handled totally differently.

Nevertheless, I don't think this constructor is the best place for this "if something is wrong, fail program softly" code. Now when I think it, the correct place would be a static function `BuildInfo.initialize()`, which would call this constructor. This constructor should throw (non-runtime) exception if version is null. The `INSTANCE` field should be initialized like `BuildInfo INSTANCE = initialize();`. So I'll change this, if you agree that that would be better.

*) The best option is perhaps the combination of the two: throw some specific runtime exception but make sure it is caught by some common runtime error handler, which writes the error log and then either recovers or terminates. But I don't know how to create such a handler with this framework, and it's so out of the scope of this task that I won't do it now. If something like that is later implemented, these code locations of "soft failures" can be easily found searching "Logger.error".